### PR TITLE
WIP: Allow user configuration of new machines

### PR DESCRIPTION
**Pull request by multilinear**
The idea is to let someone specify bash commands that run at the end of the initial setup script. They could, for example do this:
EXTRA_MINION_USER_DATA="${cat some_script_file}" 

And thus users can have some extra configuration occur each time a new minion is created.

### CODE REVIEW

**Comment by mbforbes** [comment]
@zmerlynn what do you think?

**Comment by mbforbes** [comment]
@multilinear have you signed a CLA?

**Comment by mbforbes** [comment]
(@zmerlynn for primary check; please pass off to two TLs if this looks good / for triage & review)

**Comment by multilinear** [comment]
I haven't signed the CLA just yet, sorry about that... I'll get it worked out today.

File: cluster/aws/options.md [hunk 1]
```diff
@@ -70,6 +70,20 @@ will run on this storage if available, as typically the root disk is comparative
 
 If your machines don't have any ephemeral disks, this will default to the aufs driver on your root disk (with no LVM).
 
+## EXTRA_MINION_USER_DATA
```
Comment (justinsb) [inline cluster/aws/options.md]: KUBE_EXTRA_MINION_USER_DATA, I think

File: cluster/aws/options.md [hunk 2]
```diff
@@ -70,6 +70,20 @@ will run on this storage if available, as typically the root disk is comparative
...
+
+If set to non-empty the contents of this variable will be appended to the end of the "minon-user-data" before it's
+passed to AWS. minion-user-data is a script that gets run on a new AWS instance on instance creation.
+
+Value defaults to empty string.
+
+Setting this variable to a string containing bash code allows the user to do some additional system configuration of
+every minion.
+
+## EXTRA_MASTER_USER_DATA
```
Comment (justinsb) [inline cluster/aws/options.md]: KUBE_EXTRA_MASTER_USER_DATA

File: cluster/aws/options.md [hunk 3]
```diff
@@ -70,6 +70,20 @@ will run on this storage if available, as typically the root disk is comparative
...
+
+This is the same thing as `EXTRA_MINION_USER_DATA` but for the master instead.
```
Comment (justinsb) [inline cluster/aws/options.md]: KUBE_

File: cluster/aws/coreos/util.sh [hunk 4]
```diff
@@ -51,6 +51,9 @@ function generate-minion-user-data() {
       KUBERNETES_CONTAINER_RUNTIME=$(yaml-quote ${CONTAINER_RUNTIME})
       RKT_VERSION=$(yaml-quote ${RKT_VERSION})
 EOF
+  if [[ -n "${EXTRA_MINION_USER_DATA}" ]]; then
```
Comment (justinsb) [inline cluster/aws/coreos/util.sh]: I think we're in YAML for CoreOS configuration.  I think remove this and just add a comment saying "not currently supported for CoreOS" to config-default/config-test & options.md

Comment (justinsb) [inline cluster/aws/coreos/util.sh]: Right @bakins?

File: cluster/aws/util.sh [hunk 5]
```diff
@@ -760,6 +760,9 @@ function kube-up {
     grep -v "^#" "${KUBE_ROOT}/cluster/aws/templates/create-dynamic-salt-files.sh"
     grep -v "^#" "${KUBE_ROOT}/cluster/aws/templates/download-release.sh"
     grep -v "^#" "${KUBE_ROOT}/cluster/aws/templates/salt-master.sh"
+    if [[ -n "${EXTRA_MASTER_USER_DATA}" ]]; then
+      echo "${EXTRA_MASTER_USR_DATA}"
```
Comment (justinsb) [inline cluster/aws/util.sh]: s/USR/USER/g

File: cluster/aws/options.md (hunk 3 above)
Comment (justinsb) [inline cluster/aws/options.md]: You can just document them both together and say "...for the minion/master respectively" or something like that, if that is clearer.

**Comment by justinsb** [comment]
I like this suggestion (full disclosure - I work with @multilinear at meteor).  It lets us continue to use official kube-up while adding some extra scripts we need (e.g. to do periodic rebalancing of btrfs).  We want to get that upstream long-term, but we'll always have some stuff short-term we're working on, which won't be in the last release.

**Comment by erictune** [comment]
Seems reasonable, but probably post-v1?

**Comment by zmerlynn** [comment]
This won't work on GCE, which has had a static script for a while. There's a way to make it work - get it plumbed into the `kube-env` or a separate piece of instance metadata, but as @erictune pointed out, I'm also not sure this makes the v1 cut for now.

**Comment by mikedanese** [comment]
Needs rebase and squash.

**Comment by multilinear** [comment]
Sorry, I moved on to other projects, no longer going to do this

**Pull request closed without merging**